apps/settings/js/panels/simpin_dialog/simpin_dialog.js

```diff
@@ -0,0 +1,590 @@
+          return this._elements.pinInput.value;
+
+        // unlock SIM
+        case 'unlock_pin':
+          return this._unlockPin();
+        case 'unlock_puk':
+          return this._unlockPuk('puk');
+        case 'unlock_puk2':
+          return this._unlockPuk('puk2');
+
+        // PIN lock
+        case 'enable_lock':
+          return this._enableLock(true);
+        case 'disable_lock':
+          return this._enableLock(false);
+        case 'change_pin':
+          return this._changePin('pin');
+
+        // get PIN2 code (FDN contact list)
+        case 'get_pin2':
+          return this._updateFdnContact();
+
+        // PIN2 lock (FDN)
+        case 'enable_fdn':
+          return this._enableFdn(true);
+        case 'disable_fdn':
+          return this._enableFdn(false);
+        case 'change_pin2':
+          return this._changePin('pin2');
+
+        default:
+          return Promise.reject();
+      }
+    },
+
+    /**
+     * We will clear all pre-set values when the dialog is hidden.
+     *
+     * @memberOf SimPinDialog
+     * @access public
+     */
+    clear: function() {
+      this._elements.errorMsg.hidden = true;
+      this._elements.pinInput.value = '';
+      this._elements.pukInput.value = '';
+      this._elements.newPinInput.value = '';
+      this._elements.confirmPinInput.value = '';
+    },
+
+    /**
+     * Unlock pin
+     *
+     * @memberOf SimPinDialog
+     * @access private
+     * @return {Promise}
+     */
+    _unlockPin: function() {
+      var pin = this._elements.pinInput.value;
+      if (pin === '') {
+        return Promise.reject();
+      }
+      return this._unlockCardLock({
+        lockType: 'pin',
+        pin: pin
+      });
+    },
+
+    /**
+     * Unlock puk1 or puk2
+     *
+     * @memberOf SimPinDialog
+     * @param {String} lockType
+     * @access private
+     * @return {Promise}
+     */
+    _unlockPuk: function(lockType) {
+      lockType = lockType || 'puk';
+      var puk = this._elements.pukInput.value;
+      var newPin = this._elements.newPinInput.value;
+      var confirmPin = this._elements.confirmPinInput.value;
+
+      if (puk === '' || newPin === '' || confirmPin === '') {
+        return Promise.reject();
+      }
+
+      if (newPin !== confirmPin) {
+        this._elements.showMessage('newPinErrorMsg');
+        this._elements.newPinInput.value = '';
+        this._elements.confirmPinInput.value = '';
+        return Promise.reject();
+      }
+
+      return this._unlockCardLock({
+        lockType: lockType,
+        puk: puk,
+        newPin: newPin
+      });
+    },
+
+    /**
+     * This is an internal function to pass all arguments to icc and waiting
+     * for returned results.
+     *
+     * @memberOf SimPinDialog
+     * @param {Object} options
+     * @access private
+     * @return {Promise}
+     */
+    _unlockCardLock: function(options) {
+      var defer = Defer();
+      var req = this._icc.unlockCardLock(options);
```

Comment:
Just confirmed with Edgar, the result returning from the api is a Promise (a dom request acts as a Promise), so let's use a promise chain here.